tenantcostclient: smooth over debt payback

RaduBerinde opened the pull request:
Prior to this commit, the tenant's local token bucket stalls all
traffic whenever it incurs debt. This is a problem because some debt
comes in discrete "packets" (CPU usage is accounted for once a
second), which can lead to large occasional latency spikes.

This change improves the debt repayment mechanism: instead of paying
the debt as soon as possible, we aim to pay the debt within 2 seconds
of when the debt is incurred and earmark part of the rate for debt
repayment; the remaining rate can be used by requests in the mean
time. The comment in `token_bucket.go` goes over more details.

Release note: None

Release justification: Necessary fix for the distributed rate limiting
functionality, which is vital for the upcoming Serverless MVP release.
It allows CRDB to throttle clusters that have run out of free or paid
request units (which measure CPU and I/O usage). This functionality is
only enabled in multi-tenant scenarios and should have no impact on
our dedicated customers.

### Plots for latency data from the debt test

Before:
![image](https://user-images.githubusercontent.com/16544120/136071271-fee831e8-593a-4d82-b7f7-4ac30c0e3e1c.png)


After:

![image](https://user-images.githubusercontent.com/16544120/136071244-a86afbc5-55f8-4c26-b578-bc3e31538ffd.png)

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71149)
<!-- Reviewable:end -->

andy-kimball reviewed (commented):
I'm surprised you don't have any data-driven tests here. Are those not useful for this new case?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71149#-:-MlN8-mD9e6K_DZH1jgx:b9wbidk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RaduBerinde)

___
*[pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71149#-MlN5GETBLankI4Q9FBF:-MlN5GETBLankI4Q9FBG:bpknp7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/74c7a5ce2f2035f709a80c6fa44afe385f2e4068/pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go#L48)):*
> ```Go
> // This rate is recalculated every time we incur debt. So in essence, every time
> // we incur debt, we put it together with all existing debt and plan to pay it
> // within time T (we can think of this as "refinancing" the existing debt).
> ```

:joy: We should charge an interest rate.
___
*[pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go, line 172 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/71149#-MlN6XDGDx3AISYCf6_j:-MlN6XDGDx3AISYCf6_k:bhrb3vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/74c7a5ce2f2035f709a80c6fa44afe385f2e4068/pkg/ccl/multitenantccl/tenantcostclient/token_bucket.go#L172)):*
> ```Go
> // increasing or decreasing them.
> //
> // A positive amount indicates
> ```

NIT: This comment needs fixing up.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
The datadriven tests are very hard to set up. I added one now - it took me ~1hr to get it right :)
TestTokenBucketTryToFulfill was much more useful.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71149#-:-MlNpjQa2ZdhwzMqdL8G:b-ktxfsk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andy-kimball)



<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71149#-:-MlQlxPX6ZE5xHfXj_QK:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @andy-kimball)



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71149#-:-MlQm5_B0G9UEbR3tKFw:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @andy-kimball)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
tenantcostclient: smooth over debt payback

Prior to this commit, the tenant's local token bucket stalls all
traffic whenever it incurs debt. This is a problem because some debt
comes in discrete "packets" (CPU usage is accounted for once a
second), which can lead to large occasional latency spikes.

This change improves the debt repayment mechanism: instead of paying
the debt as soon as possible, we aim to pay the debt within 2 seconds
of when the debt is incurred and earmark part of the rate for debt
repayment; the remaining rate can be used by requests in the mean
time. The comment in `token_bucket.go` goes over more details.

Release note: None

Release justification: Necessary fix for the distributed rate limiting
functionality, which is vital for the upcoming Serverless MVP release.
It allows CRDB to throttle clusters that have run out of free or paid
request units (which measure CPU and I/O usage). This functionality is
only enabled in multi-tenant scenarios and should have no impact on
our dedicated customers.

RaduBerinde commented:
Ah, the problem was that I wasn't rebased on top of the latest cost model changes. Fixed, will try again after CI runs.

RaduBerinde commented:
bors r+
